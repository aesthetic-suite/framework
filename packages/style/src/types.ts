import {
  ClassName,
  CSS,
  Direction,
  GenericProperties,
  Properties,
  Property,
  Unit,
  UnitFactory,
  Value,
  Variables,
} from '@aesthetic/types';

export type SheetType = 'global' | 'standard' | 'conditions';

export type Condition = string;

export type RankCache = Record<string, number>;

export interface ProcessOptions {
  deterministic?: boolean;
  direction?: Direction;
  rankings?: RankCache;
  type?: SheetType;
  unit?: Unit | UnitFactory;
  vendor?: boolean;
}

export interface RenderOptions extends ProcessOptions {
  className?: ClassName;
  conditions?: Condition[] | string;
  // minimumRank?: number;
  selector?: string;
}

export interface CacheItem {
  className: string;
  rank?: number;
}

export interface StyleRule {
  conditionText?: string;
  cssRules: StyleRule[];
  cssText: CSS;
  cssVariables: Variables<string>;
  textContent: CSS;
  type: number;
  insertRule: (rule: CSS, index: number) => number;
}

// ADDONS

export interface CacheManager {
  read: (key: string, minimumRank?: number) => CacheItem | null;
  write: (key: string, item: CacheItem) => void;
}

export interface DirectionConverter {
  convert: <T extends Value>(
    from: Direction,
    to: Direction,
    property: string,
    value: T,
  ) => { property: string; value: T };
}

export interface SheetManager {
  insertRule: (type: SheetType, rule: CSS, index?: number) => number;
}

export interface VendorPrefixer {
  prefix: (key: string, value: Value) => GenericProperties;
  prefixSelector: (selector: string, rule: CSS) => CSS;
}

export interface API {
  converter?: DirectionConverter | null;
  direction: Direction;
  prefixer?: VendorPrefixer | null;
}

export interface EngineOptions {
  cacheManager: CacheManager;
  direction: Direction;
  directionConverter?: DirectionConverter | null;
  sheetManager: SheetManager;
  vendorPrefixer?: VendorPrefixer | null;
}

export interface Engine {
  renderDeclaration: <K extends Property>(
    property: K,
    value: Properties[K],
    options?: RenderOptions,
  ) => ClassName;
}
