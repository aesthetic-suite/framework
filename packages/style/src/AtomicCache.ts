import { CacheItem, StyleParams } from './types';

// https://jsperf.com/javascript-objects-vs-map-performance
// https://jsperf.com/performance-of-map-vs-object
export default class AtomicCache {
  cache: { [property: string]: { [value: string]: CacheItem[] } } = {};

  match(item: CacheItem, params: StyleParams, minimumRank?: number): boolean {
    if (minimumRank !== undefined && item.rank < minimumRank) {
      return false;
    }

    if (item.selector !== params.selector) {
      return false;
    }

    if (item.type !== params.type) {
      return false;
    }

    if (Array.isArray(params.conditions) && params.conditions.length > 0) {
      return params.conditions.every(i =>
        item.conditions.find(p => p.query === i.query && p.type === i.type),
      );
    }

    return true;
  }

  read(
    property: string,
    value: string,
    params: StyleParams,
    minimumRank?: number,
  ): CacheItem | null {
    const propertyCache = this.cache[property];

    if (!propertyCache) {
      return null;
    }

    const valueCache = propertyCache[value];

    if (!valueCache || valueCache.length === 0) {
      return null;
    }

    return valueCache.find(item => this.match(item, params, minimumRank)) ?? null;
  }

  write(property: string, value: string, item: CacheItem): this {
    let propertyCache = this.cache[property];

    if (!propertyCache) {
      propertyCache = {};
      this.cache[property] = propertyCache;
    }

    let valueCache = propertyCache[value];

    if (!valueCache) {
      valueCache = [];
      propertyCache[value] = valueCache;
    }

    valueCache.push(item);

    return this;
  }
}
