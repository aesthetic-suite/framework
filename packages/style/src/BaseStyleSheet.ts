import { isSSR } from '@aesthetic/utils';
import { StyleRule } from './types';

export default abstract class BaseStyleSheet {
  lastIndex: number = -1;

  sheet: StyleRule;

  protected rafHandle: number = 0;

  protected ruleBuffer: { index: number; rule: string }[] = [];

  constructor(sheet: StyleRule) {
    this.sheet = sheet;
  }

  protected enqueueRule(rule: string, customIndex?: number): number {
    let index = 0;

    // When testing or SSR, just insert the rule immediately
    if (process.env.NODE_ENV === 'test' || isSSR()) {
      index = customIndex ?? this.sheet.cssRules.length;

      this.injectRule(rule, index);

      // Otherwise buffer the insert until the next animation frame
    } else {
      index = customIndex ?? (this.lastIndex += 1);

      this.ruleBuffer.push({
        index,
        rule,
      });

      if (!this.rafHandle) {
        this.rafHandle = requestAnimationFrame(this.flushRules);
      }
    }

    return index;
  }

  protected injectRule = (rule: string, index: number) => {
    // CSS is locked in developer tools when using `insertRule`,
    // so avoid using it in development so we have full control.
    // istanbul ignore next
    if (process.env.NODE_ENV === 'development' && !isSSR()) {
      this.sheet.textContent += rule;
    } else {
      this.sheet.insertRule(rule, index);
      this.lastIndex = this.sheet.cssRules.length;
    }
  };

  protected flushRules = () => {
    this.rafHandle = 0;
    this.ruleBuffer.forEach((buffer) => {
      this.injectRule(buffer.rule, buffer.index);
    });
  };
}
