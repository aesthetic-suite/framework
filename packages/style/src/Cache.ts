import { Value, ValueWithFallbacks } from '@aesthetic/types';
import { CacheManager, CacheItem, CacheStorage, RenderOptions } from './types';

export function createCacheKey(
  property: string,
  value: Value | ValueWithFallbacks,
  { selector, conditions }: RenderOptions,
): string {
  let key = `${property}:${value}`;

  if (selector) {
    key += `:${selector}`;
  }

  if (conditions && conditions.length > 0) {
    key += `:${conditions.join(':')}`;
  }

  return key;
}

export function createCacheManager(defaultItems: CacheStorage = {}): CacheManager {
  const cache: CacheStorage = defaultItems;

  return {
    read(key, minimumRank) {
      const items = cache[key];

      if (!items) {
        return null;
      } else if (minimumRank === undefined) {
        return items[0];
      }

      return items.find((item) => item.rank! >= minimumRank) || null;
    },

    write(key: string, item: CacheItem) {
      const result = cache[key] || [];

      result.push(item);

      cache[key] = result;
    },
  };
}
