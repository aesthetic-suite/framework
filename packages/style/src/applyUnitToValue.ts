import { isUnitlessProperty } from 'css-in-js-utils';
import { Value } from './types';

export default function applyUnitToValue(property: string, value: Value): string {
  if (typeof value === 'string' || isUnitlessProperty(property) || value === 0) {
    return String(value);
  }

  return `${value}px`;
}
