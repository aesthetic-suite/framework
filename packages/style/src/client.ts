import { Variables } from '@aesthetic/types';
import { objectLoop } from '@aesthetic/utils';
import createStyleElements, { getStyleElement } from './client/dom';
import hydrateStyles from './client/hydration';
// Rollup compatibility
import {
  createCacheManager,
  createStyleEngine,
  createSheetManager,
  formatVariable,
  EngineOptions,
  StyleEngine,
} from './index';

export { createStyleElements, getStyleElement };

function setRootVariables(vars: Variables) {
  const root = document.documentElement;

  objectLoop(vars, (value, key) => {
    root.style.setProperty(formatVariable(key), String(value));
  });
}

export function createClientEngine(options: Partial<EngineOptions> = {}): StyleEngine {
  const engine: StyleEngine = {
    ...createStyleEngine({
      cacheManager: createCacheManager(),
      sheetManager: createSheetManager(createStyleElements()),
      ...options,
    }),
    setRootVariables,
  };

  // Attempt to hydrate styles immediately
  hydrateStyles(engine);

  return engine;
}
