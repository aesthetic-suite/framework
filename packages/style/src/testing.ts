/* eslint-disable unicorn/import-index */

import { CSS } from '@aesthetic/types';
import {
  CacheManager,
  StyleEngine,
  EngineOptions,
  SheetManager,
  SheetType,
  StyleRule,
  getDocumentStyleSheet,
} from './index';
import { createCacheManager, createStyleElements, createEngine, createSheetManager } from './next';

export function createTestCacheManager(): CacheManager {
  return createCacheManager();
}

export function createTestSheetManager(): SheetManager {
  return createSheetManager(createStyleElements());
}

export function createTestStyleEngine(options: Partial<EngineOptions>): StyleEngine {
  return createEngine({
    cacheManager: createTestCacheManager(),
    sheetManager: createTestSheetManager(),
    ...options,
  });
}

export function getRenderedStyles(type: SheetType | StyleRule): CSS {
  return Array.from(
    ((typeof type === 'string' ? getDocumentStyleSheet(type) : type) as StyleRule).cssRules,
  ).reduce((css, rule) => css + rule.cssText, '');
}

export function purgeStyles(type?: SheetType): void {
  if (type) {
    // This is the only way to generate accurate snapshots.
    // It may slow down tests though?
    document.getElementById(`aesthetic-${type}`)?.remove();
  } else {
    purgeStyles('global');
    purgeStyles('standard');
    purgeStyles('conditions');
  }
}
