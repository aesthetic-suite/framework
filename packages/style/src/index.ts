/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { Direction, EngineOptions, VariablesMap } from '@aesthetic/types';
import { objectLoop } from '@aesthetic/utils';
import { createStyleElements, getStyleElement } from './client/dom';
import { hydrateStyles } from './client/hydration';
import { createCacheManager } from './common/cache';
import { createStyleEngine } from './common/engine';
import { createSheetManager } from './common/sheet';
import { formatVariable } from './common/syntax';
import { StyleEngine } from './types';

export * from './common/cache';
export * from './common/constants';
export * from './common/engine';
export * from './common/helpers';
export * from './common/sheet';
export * from './common/syntax';
export * from './types';

export { createStyleElements, getStyleElement };

// Set active direction on the document `dir` attribute
function setDirection(direction: Direction) {
  document.documentElement.setAttribute('dir', direction);
}

// Set CSS variables to :root
function setRootVariables(vars: VariablesMap) {
  objectLoop(vars, (value, key) => {
    document.documentElement.style.setProperty(formatVariable(key), String(value));
  });
}

// Set active theme class names on the `body`
function setTheme(classNames: string[]) {
  document.body.className = classNames.join(' ');
}

export function createClientEngine(options: Partial<EngineOptions> = {}): StyleEngine {
  if (global.AESTHETIC_CUSTOM_ENGINE) {
    return global.AESTHETIC_CUSTOM_ENGINE as StyleEngine;
  }

  const direction = (document.documentElement.getAttribute('dir') ||
    document.body.getAttribute('dir') ||
    'ltr') as Direction;

  const engine: StyleEngine = createStyleEngine({
    cacheManager: createCacheManager(),
    direction,
    sheetManager: createSheetManager(createStyleElements()),
    ...options,
  });

  // Match against browser preferences
  engine.prefersColorScheme = (scheme) => matchMedia(`(prefers-color-scheme: ${scheme})`).matches;
  engine.prefersContrastLevel = (level) => matchMedia(`(prefers-contrast: ${level})`).matches;

  // Handle DOM specific logic
  engine.setDirection = setDirection;
  engine.setRootVariables = setRootVariables;
  engine.setTheme = setTheme;

  // Attempt to hydrate styles immediately
  hydrateStyles(engine);

  return engine;
}
