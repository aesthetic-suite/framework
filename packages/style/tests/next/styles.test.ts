import directionConverter from '@aesthetic/addon-direction';
import vendorPrefixer from '@aesthetic/addon-vendor';
import { createTestStyleEngine, getRenderedStyles, purgeStyles } from '../../src/testing';
import { Engine } from '../../src/types';

describe('Engine', () => {
  let engine: Engine;

  beforeEach(() => {
    engine = createTestStyleEngine({
      directionConverter,
      vendorPrefixer,
    });
  });

  afterEach(() => {
    purgeStyles();
  });

  describe('renderDeclaration()', () => {
    it('generates a unique class name for a large number of properties', () => {
      for (let i = 0; i < 100; i += 1) {
        engine.renderDeclaration('padding', `${i}px`);
      }

      expect(getRenderedStyles('standard')).toMatchSnapshot();
    });

    it('uses the same class name for the same property value pair', () => {
      engine.renderDeclaration('display', 'block');
      engine.renderDeclaration('display', 'flex');
      engine.renderDeclaration('display', 'block');
      engine.renderDeclaration('display', 'flex');
      engine.renderDeclaration('display', 'inline');
      engine.renderDeclaration('display', 'block');

      expect(getRenderedStyles('standard')).toMatchSnapshot();
    });

    it('uses the same class name for dashed and camel cased properties', () => {
      engine.renderDeclaration('textDecoration', 'none');
      engine.renderDeclaration('text-decoration', 'none');

      expect(getRenderedStyles('standard')).toMatchSnapshot();
    });

    it('uses the same class name for numeric and string values', () => {
      engine.renderDeclaration('width', 0);
      engine.renderDeclaration('width', '0');
      engine.renderDeclaration('width', '100em');

      expect(getRenderedStyles('standard')).toMatchSnapshot();
    });

    it('supports CSS variables', () => {
      engine.renderDeclaration('color', 'var(--primary-color)');
      engine.renderDeclaration('border', '1px solid var(--border-color)');
      engine.renderDeclaration('display', 'var(--display, var(--fallback), flex)');

      expect(getRenderedStyles('standard')).toMatchSnapshot();
    });

    it('supports fallback values by using an array', () => {
      engine.renderDeclaration('display', ['box', 'block', 'flex']);

      // Wont show in snapshot accurately
      expect(getRenderedStyles('standard')).toMatchSnapshot();
    });

    it('supports selectors', () => {
      engine.renderDeclaration('color', 'green', { selector: ':hover' });
      engine.renderDeclaration('color', 'red', { selector: '[disabled]' });
      engine.renderDeclaration('color', 'blue', { selector: ':nth-child(2)' });

      expect(getRenderedStyles('standard')).toMatchSnapshot();
    });

    it('supports conditionals', () => {
      engine.renderDeclaration('color', 'green', { conditions: ['@media (max-size: 100px)'] });
      engine.renderDeclaration('color', 'red', { conditions: ['@supports (color: red)'] });
      engine.renderDeclaration('color', 'blue', {
        conditions: [
          '@media (max-width: 100px)',
          '@supports (color: red)',
          '@media (min-width: 200px)',
        ],
      });

      expect(getRenderedStyles('conditions')).toMatchSnapshot();
    });

    it('supports conditionals using strings', () => {
      engine.renderDeclaration('color', 'green', {
        conditions: '@media (max-size: 100px) { #rule# }',
      });
      engine.renderDeclaration('color', 'red', { conditions: '@supports (color: red) { #rule# }' });
      engine.renderDeclaration('color', 'blue', {
        conditions:
          '@media (max-width: 100px) { @supports (color: red) { @media (min-width: 200px) { #rule# } } }',
      });

      expect(getRenderedStyles('conditions')).toMatchSnapshot();
    });

    it('supports conditionals with selectors', () => {
      engine.renderDeclaration('color', 'green', {
        conditions: ['@media (max-size: 100px)'],
        selector: ':focus',
      });

      expect(getRenderedStyles('conditions')).toMatchSnapshot();
    });

    it('converts directional properties', () => {
      engine.renderDeclaration('margin-left', 0);
      engine.renderDeclaration('margin-right', 0);
      engine.renderDeclaration('margin-left', 0, { direction: 'ltr' });
      engine.renderDeclaration('margin-right', 0, { direction: 'rtl' });
      engine.renderDeclaration('margin-left', 0, { direction: 'ltr' });
      engine.renderDeclaration('margin-right', 0, { direction: 'rtl' });

      expect(getRenderedStyles('standard')).toMatchSnapshot();
    });

    it('converts directional values', () => {
      engine.renderDeclaration('text-align', 'left');
      engine.renderDeclaration('text-align', 'right');
      engine.renderDeclaration('text-align', 'left', { direction: 'ltr' });
      engine.renderDeclaration('text-align', 'left', { direction: 'rtl' });
      engine.renderDeclaration('text-align', 'right', { direction: 'ltr' });
      engine.renderDeclaration('text-align', 'right', { direction: 'rtl' });

      expect(getRenderedStyles('standard')).toMatchSnapshot();
    });

    it('prefixes properties, values, value functions, and selectors', () => {
      // Value prefixing (wont show in snapshot because of DOM)
      engine.renderDeclaration('min-width', 'fit-content', { vendor: true });

      // Value function prefixing (wont show in snapshot because of DOM)
      engine.renderDeclaration('background', 'image-set()', { vendor: true });

      // Property prefixing
      engine.renderDeclaration('appearance', 'none', { vendor: true });

      // Selector prefixing
      engine.renderDeclaration('display', 'none', {
        selector: ':fullscreen',
        vendor: true,
      });

      expect(getRenderedStyles('standard')).toMatchSnapshot();
    });
  });
});
