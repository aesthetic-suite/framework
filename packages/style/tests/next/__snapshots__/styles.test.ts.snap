// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Engine renderDeclaration() converts directional properties 1`] = `".a {margin-left: 0;}.b {margin-right: 0;}"`;

exports[`Engine renderDeclaration() converts directional values 1`] = `".a {text-align: left;}.b {text-align: right;}"`;

exports[`Engine renderDeclaration() generates a unique class name for a large number of properties 1`] = `".a {padding: 0px;}.b {padding: 1px;}.c {padding: 2px;}.d {padding: 3px;}.e {padding: 4px;}.f {padding: 5px;}.g {padding: 6px;}.h {padding: 7px;}.i {padding: 8px;}.j {padding: 9px;}.k {padding: 10px;}.l {padding: 11px;}.m {padding: 12px;}.n {padding: 13px;}.o {padding: 14px;}.p {padding: 15px;}.q {padding: 16px;}.r {padding: 17px;}.s {padding: 18px;}.t {padding: 19px;}.u {padding: 20px;}.v {padding: 21px;}.w {padding: 22px;}.x {padding: 23px;}.y {padding: 24px;}.z {padding: 25px;}.a1 {padding: 26px;}.b1 {padding: 27px;}.c1 {padding: 28px;}.d1 {padding: 29px;}.e1 {padding: 30px;}.f1 {padding: 31px;}.g1 {padding: 32px;}.h1 {padding: 33px;}.i1 {padding: 34px;}.j1 {padding: 35px;}.k1 {padding: 36px;}.l1 {padding: 37px;}.m1 {padding: 38px;}.n1 {padding: 39px;}.o1 {padding: 40px;}.p1 {padding: 41px;}.q1 {padding: 42px;}.r1 {padding: 43px;}.s1 {padding: 44px;}.t1 {padding: 45px;}.u1 {padding: 46px;}.v1 {padding: 47px;}.w1 {padding: 48px;}.x1 {padding: 49px;}.y1 {padding: 50px;}.z1 {padding: 51px;}.a2 {padding: 52px;}.b2 {padding: 53px;}.c2 {padding: 54px;}.d2 {padding: 55px;}.e2 {padding: 56px;}.f2 {padding: 57px;}.g2 {padding: 58px;}.h2 {padding: 59px;}.i2 {padding: 60px;}.j2 {padding: 61px;}.k2 {padding: 62px;}.l2 {padding: 63px;}.m2 {padding: 64px;}.n2 {padding: 65px;}.o2 {padding: 66px;}.p2 {padding: 67px;}.q2 {padding: 68px;}.r2 {padding: 69px;}.s2 {padding: 70px;}.t2 {padding: 71px;}.u2 {padding: 72px;}.v2 {padding: 73px;}.w2 {padding: 74px;}.x2 {padding: 75px;}.y2 {padding: 76px;}.z2 {padding: 77px;}.a3 {padding: 78px;}.b3 {padding: 79px;}.c3 {padding: 80px;}.d3 {padding: 81px;}.e3 {padding: 82px;}.f3 {padding: 83px;}.g3 {padding: 84px;}.h3 {padding: 85px;}.i3 {padding: 86px;}.j3 {padding: 87px;}.k3 {padding: 88px;}.l3 {padding: 89px;}.m3 {padding: 90px;}.n3 {padding: 91px;}.o3 {padding: 92px;}.p3 {padding: 93px;}.q3 {padding: 94px;}.r3 {padding: 95px;}.s3 {padding: 96px;}.t3 {padding: 97px;}.u3 {padding: 98px;}.v3 {padding: 99px;}"`;

exports[`Engine renderDeclaration() prefixes properties, values, value functions, and selectors 1`] = `".a {min-width: fit-content;}.b {background: image-set();}.c {appearance: none; -moz-appearance: none; -webkit-appearance: none;}.d:-ms-fullscreen {display: none;}"`;

exports[`Engine renderDeclaration() supports CSS variables 1`] = `".a {color: var(--primary-color);}.b {border: 1px solid var(--border-color);}.c {display: var(--display, var(--fallback), flex);}"`;

exports[`Engine renderDeclaration() supports conditionals 1`] = `"@media (max-size: 100px) {.a {color: green;}}@supports (color: red) {.b {color: red;}}@media (max-width: 100px) {@supports (color: red) {@media (min-width: 200px) {.c {color: blue;}}}}"`;

exports[`Engine renderDeclaration() supports conditionals using strings 1`] = `"@media (max-size: 100px) {.a {color: green;}}@supports (color: red) {.b {color: red;}}@media (max-width: 100px) {@supports (color: red) {@media (min-width: 200px) {.c {color: blue;}}}}"`;

exports[`Engine renderDeclaration() supports conditionals with selectors 1`] = `"@media (max-size: 100px) {.a:focus {color: green;}}"`;

exports[`Engine renderDeclaration() supports fallback values by using an array 1`] = `".a {display: flex;}"`;

exports[`Engine renderDeclaration() supports selectors 1`] = `".a:hover {color: green;}.b[disabled] {color: red;}.c:nth-child(2) {color: blue;}"`;

exports[`Engine renderDeclaration() uses the same class name for dashed and camel cased properties 1`] = `".a {text-decoration: none;}"`;

exports[`Engine renderDeclaration() uses the same class name for numeric and string values 1`] = `".a {width: 0;}.b {width: 100em;}"`;

exports[`Engine renderDeclaration() uses the same class name for the same property value pair 1`] = `".a {display: block;}.b {display: flex;}.c {display: inline;}"`;
